src/diag/map_editor.cpp

```diff
@@ -5,149 +5,96 @@
-            return S_PETROL_UPSHIFT_MAP;
-        case MAP_ID_S_PETROL_DOWNSHIFT:
-            return S_PETROL_DOWNSHIFT_MAP;
-        case MAP_ID_C_DIESEL_UPSHIFT:
-            return C_DIESEL_UPSHIFT_MAP;
-        case MAP_ID_C_DIESEL_DOWNSHIFT:
-            return C_DIESEL_DOWNSHIFT_MAP;
-        case MAP_ID_C_PETROL_UPSHIFT:
-            return C_PETROL_UPSHIFT_MAP;
-        case MAP_ID_C_PETROL_DOWNSHIFT:
-            return C_PETROL_DOWNSHIFT_MAP;
-        case MAP_ID_A_DIESEL_UPSHIFT:
-            return A_DIESEL_UPSHIFT_MAP;
-        case MAP_ID_A_DIESEL_DOWNSHIFT:
-            return A_DIESEL_DOWNSHIFT_MAP;
-        case MAP_ID_A_PETROL_UPSHIFT:
-            return A_PETROL_UPSHIFT_MAP;
-        case MAP_ID_A_PETROL_DOWNSHIFT:
-            return A_PETROL_DOWNSHIFT_MAP;
-        default:
-            return nullptr;
+#define CHECK_MAP(map_id) \
+    StoredTcuMap* ptr = get_map(map_id); \
+    if (ptr == nullptr) { \
+        return NRC_SUB_FUNC_NOT_SUPPORTED_INVALID_FORMAT; \
     }
-}
 
-uint8_t MapEditor::read_map_data(uint8_t map_id, uint16_t *dest_size, int16_t** buffer) {
-    uint8_t map_idx = map_id & 0b1111111;
-    bool is_default_map = (map_id & 0b10000000) != 0;
-    if (map_idx >= 0x01 && map_idx <= 0x0C) {
-        // Shift map
-        int16_t* b = (int16_t*)heap_caps_malloc(SHIFT_MAP_SIZE*sizeof(int16_t), MALLOC_CAP_SPIRAM);
-        if (b == nullptr) {
-            ESP_LOGE("MAP_EDITOR_R", "Could not allocate read array!");
-            return NRC_GENERAL_REJECT;
-        }
-        if (is_default_map) {
-            // Read default map from flash
-            const int16_t* default_map = get_default_map(map_idx);
-            if (default_map == nullptr) {
-                ESP_LOGE("MAP_EDITOR_R", "default map is invalid!?");
-                delete b;
-                return NRC_GENERAL_REJECT;
-            }
-            memcpy(b, default_map, SHIFT_MAP_SIZE*sizeof(int16_t));
-            *buffer = b;
-            *dest_size = SHIFT_MAP_SIZE*sizeof(int16_t);
-            return 0;
-        } else {
-            const char* name = map_id_to_name(map_idx);
-            if (name == nullptr) {
-                ESP_LOGE("MAP_EDITOR_R", "map name is null!?");
-                delete b;
-                return NRC_GENERAL_REJECT;
-            }
-            if (EEPROM::read_nvs_map_data(name, b, nullptr, SHIFT_MAP_SIZE)) {
-                // OK, copy the pointer
-                *buffer = b;
-                *dest_size = SHIFT_MAP_SIZE*sizeof(int16_t);
-                return 0;
-            } else {
-                // De-allocate
-                ESP_LOGE("MAP_EDITOR_R", "read_nvs_map_data failed!");
-                delete b;
-                return NRC_GENERAL_REJECT;
-            }
-        }
+uint8_t MapEditor::read_map_data(uint8_t map_id, bool is_default, uint16_t *dest_size, int16_t** buffer) {
+    CHECK_MAP(map_id)
+    // Map valid
+    int size = ptr->get_map_element_count();
```

Comment:
Consider using an unsigned integer. "size" is only used in combination with other unsigned integers, as far, as I can see. No need for casts here.